File: modern/_1.15.2/src/main/java/cn/enaium/noexpensive/callback/impl/EnchantmentCanCombineCallbackImpl.java
package cn.enaium.noexpensive.callback.impl;

import cn.enaium.noexpensive.Config;
import cn.enaium.noexpensive.callback.EnchantmentCanCombineCallback;
import net.minecraft.enchantment.Enchantment;
import net.minecraft.util.ActionResult;
import net.minecraft.util.registry.Registry;

import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * @author Enaium
 */
public class EnchantmentCanCombineCallbackImpl implements EnchantmentCanCombineCallback {
    @Override
    public ActionResult interact(Enchantment enchantment1, Enchantment enchantment2) {
        final String enchantment1Name = Objects.requireNonNull(Registry.ENCHANTMENT.getId(enchantment1)).toString();
        final String enchantment2Name = Objects.requireNonNull(Registry.ENCHANTMENT.getId(enchantment2)).toString();

        final Map<String, List<String>> compatibility = Config.getModel().compatibility;

        if (compatibility.containsKey(enchantment1Name) && compatibility.get(enchantment1Name).contains(enchantment2Name)) {
            return ActionResult.PASS;
        } else if (compatibility.containsKey(enchantment2Name) && compatibility.get(enchantment2Name).contains(enchantment1Name)) {
            return ActionResult.PASS;
        }
        return enchantment1.isDifferent(enchantment2) ? ActionResult.PASS : ActionResult.FAIL;
    }
}

File: modern/_1.15.2/src/main/kotlin/cn/enaium/noexpensive/callback/impl/EnchantmentCanCombineCallbackImpl.kt
package cn.enaium.noexpensive.callback.impl

import cn.enaium.noexpensive.Config
import cn.enaium.noexpensive.callback.EnchantmentCanCombineCallback
import net.minecraft.enchantment.Enchantment
import net.minecraft.util.ActionResult
import net.minecraft.util.registry.Registry
import java.util.*

/**
 * @author Enaium
 */
class EnchantmentCanCombineCallbackImpl : EnchantmentCanCombineCallback {
    override fun interact(enchantment1: Enchantment, enchantment2: Enchantment): ActionResult {
        val enchantment1Name = Objects.requireNonNull(Registry.ENCHANTMENT.getId(enchantment1)).toString()
        val enchantment2Name = Objects.requireNonNull(Registry.ENCHANTMENT.getId(enchantment2)).toString()
        val compatibility: Map<String, List<String>> = Config.model.compatibility
        if (compatibility.containsKey(enchantment1Name) && compatibility[enchantment1Name]!!.contains(enchantment2Name)) {
            return ActionResult.PASS
        } else if (compatibility.containsKey(enchantment2Name) && compatibility[enchantment2Name]!!.contains(
                enchantment1Name
            )
        ) {
            return ActionResult.PASS
        }
        return if (enchantment1.isDifferent(enchantment2)) ActionResult.PASS else ActionResult.FAIL
    }
}
